homeassistant/components/google_assistant/trait.py

```diff
@@ -348,6 +350,7 @@ def supported(domain, features, device_class, _):
             light.DOMAIN,
             media_player.DOMAIN,
             humidifier.DOMAIN,
+            water_heater.DOMAIN,
```

Comment:
A `water_heater` entity does not support turn_on and turn_off. The OnOffTrait trait is required for a WATERHEATER though by Google. How do you want to solve this part?

https://developers.home.google.com/cloud-to-cloud/guides/waterheater

Reply:
A good question that I haven't found a decent answer for.

One possibility would be to use it to control operation mode, letting off set to off and on set to the last used operation mode (or some default). But too me this seems confusing and prone to issues.

Probably the better option would be to implement the [on/off as query only](https://developers.home.google.com/cloud-to-cloud/traits/onoff#examples), and let it report the state as off unless a mode is set. That way the UI will report the state accurately, but you won't be able to change it.
If we wanted extra points we could implement the modes trait for water heater as well, and let that control the operation mode (if supported).

Reply:
Climate has a turn on/off service. Water heater misses that, perhaps it is worth to open an architectural discussion and suggest adding this service to the core water_heater entity

Reply:
Changing the core entity sounds like the best long term solution, [and there's discussion about water heater modes already](https://github.com/home-assistant/architecture/discussions/589). Maybe a change to add preset modes and implement the turn on/off service to bring it closer to climate?

I'll draft up a post and see if anyone bites

Reply:
https://github.com/home-assistant/core/pull/94817 Implements a `turn_on` / `turn off` service for water_heater. The architecture discussion was positive on adding the this. Once merged, we can go on.

Reply:
I was swayed while reading architectural discussions into broadening the scope of the question to whether a [generic temperature control entity would be viable](https://github.com/home-assistant/architecture/discussions/923).

It hasn't gained much traction and I'm not sure if it's a feasible change to make, so I think just implementing the `turn_on` and `turn_off` as you have done is the right solution. Thank you for getting it through

Now that it's been merged I'll pull in the changes and see what needs to be tweaked.

Reply:
@hookedonunix , I am working on a PR to add MQTT support for power on/off.

Reply:
@jbouwh Is that for the water heater or a generic on/off?

Reply:
https://github.com/home-assistant/core/pull/97197. It will only support pessimistic mode (no state topic), as the steeds needs to be set through the mode_state topic.